src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+/// ```
+#[derive(Serialize, Debug, Builder)]
+#[builder(
+    custom_constructor,
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct Batch {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    parent_bid: Option<String>,
+
+    /// Batch description for Faktory WEB UI.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// On success callback.
+    ///
+    /// This job will be queued by the Faktory server provided
+    /// all the jobs belonging to this batch have been executed successfully.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) success: Option<Job>,
+
+    /// On complete callback.
+    ///
+    /// This job will be queue by the Faktory server after all the jobs
+    /// belonging to this batch have been executed, even if one/some/all
+    /// of the workers have failed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder::new(description)
+    }
+}
+
+impl BatchBuilder {
+    fn build(&mut self) -> Batch {
+        self.try_build()
+            .expect("All required fields have been set.")
+    }
+
+    /// Create a new `BatchBuilder` with optional description of the batch.
+    pub fn new(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder {
+            description: Some(description.into()),
+            ..Self::create_empty()
+        }
+    }
+
+    fn success(&mut self, success_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.success = Some(success_cb.into());
+        self
+    }
+
+    fn complete(&mut self, complete_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.complete = Some(complete_cb.into());
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(&mut self, success_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with only `complete` callback specified.
+    pub fn with_complete_callback(&mut self, complete_cb: Job) -> Batch {
+        self.complete(complete_cb);
+        self.success(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(&mut self, success_cb: Job, complete_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(complete_cb);
+        self.build()
+    }
+}
+
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
```

Comment:
We should really (in the next major version) introduce dedicated type for `JobId` and `BatchId` so that folks won't accidentally mix them up as they're both strings.

Reply:
Looks like swithing to async is  perfect timing. Added to [TODOs](https://github.com/jonhoo/faktory-rs/pull/49#issue-2086696629)